### PR TITLE
[Enhancement](hudi) support hudi mor table, step2 follow #19909

### PR DESCRIPTION
**Pull request by AshinGau**
## Proposed changes
PR(https://github.com/apache/doris/pull/19909) has implemented the framework of hudi reader for MOR table. This PR completes all functions of reading MOR table and enables end-to-end queries.
Key Implementations:
1. Use hudi meta information to generate the table schema, not from hive client.
2. Use hive client to list hudi partitions, so it strongly depends the sync-tools(https://hudi.apache.org/docs/syncing_metastore/) which syncs the partitions of hudi into hive metastore. However, we may get the hudi partitions directly from .hoodie directory.
3. Remove `HudiHMSExternalCatalog`, because other catalogs like glue is compatible with hive catalog.
4. Read the COW table originally from c++.
5. Hudi RecordReader will use ProcessBuilder to start a hotspot debugger process, which may be stuck when attaching the origin JNI process, soI use a tricky method to kill this useless process.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/planner/external/hudi/HudiScanNode.java [hunk 1]
```diff
@@ -79,19 +76,24 @@ public class HudiScanNode extends HiveScanNode {
      */
     public HudiScanNode(PlanNodeId id, TupleDescriptor desc, boolean needCheckColumnPriv) {
         super(id, desc, "HUDI_SCAN_NODE", StatisticalType.HUDI_SCAN_NODE, needCheckColumnPriv);
+        isCowTable = hmsTable.isHoodieCowTable();
+        if (isCowTable) {
+            LOG.info("Hudi table {} can read as cow table", hmsTable.getName());
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/planner/external/hudi/HudiScanNode.java:81]: use debug level

File: fe/fe-core/src/main/java/org/apache/doris/planner/external/hudi/HudiScanNode.java [hunk 2]
```diff
@@ -79,19 +76,24 @@ public class HudiScanNode extends HiveScanNode {
...
+        } else {
+            LOG.info("Hudi table {} is a mor table, and will use JNI to read data in BE", hmsTable.getName());
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/planner/external/hudi/HudiScanNode.java:83]: ditto

File: fe/java-udf/pom.xml [hunk 3]
```diff
@@ -125,7 +120,7 @@ under the License.
         </dependency>
         <dependency>
             <groupId>org.apache.hudi</groupId>
-            <artifactId>hudi-presto-bundle</artifactId>
+            <artifactId>hudi-hadoop-mr-bundle</artifactId>
```
Comment (morningman) [inline fe/java-udf/pom.xml:123]: which one should be use? hadoop or presto?

File: fe/java-udf/pom.xml [hunk 4]
```diff
@@ -179,6 +168,12 @@ under the License.
             </exclusions>
         </dependency>
 
+        <dependency>
```
Comment (morningman) [inline fe/java-udf/pom.xml:171]: Why move this to the last?
If there is special reason, add comment in code.

File: fe/java-udf/src/main/java/org/apache/doris/hudi/HudiJniScanner.java [hunk 5]
```diff
@@ -31,32 +32,40 @@
...
 
+    UserGroupInformation ugi = null;
     private RecordReader<NullWritable, ArrayWritable> reader;
     private StructObjectInspector rowInspector;
     private Deserializer deserializer;
     private final ClassLoader classLoader;
 
     public HudiJniScanner(int fetchSize, Map<String, String> params) {
-        LOG.debug("fetchSize:" + fetchSize + " , params: " + params);
+        LOG.info("Hudi JNI params:\n" + params.entrySet().stream().map(kv -> kv.getKey() + "=" + kv.getValue())
```
Comment (morningman) [inline fe/java-udf/src/main/java/org/apache/doris/hudi/HudiJniScanner.java:67]: debug

File: fe/java-udf/src/main/java/org/apache/doris/jni/utils/Utils.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.lang.management.ManagementFactory;
+import java.lang.management.RuntimeMXBean;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.LinkedList;
+import java.util.List;
+
+public class Utils {
```
Comment (morningman) [inline fe/java-udf/src/main/java/org/apache/doris/jni/utils/Utils.java:36]: How about move this class to `fe-common`?

File: fe/java-udf/src/main/java/org/apache/doris/jni/vec/ColumnType.java [hunk 7]
```diff
@@ -230,6 +238,7 @@ private static int findNextNestedField(String commaSplitFields) {
     }
 
     public static ColumnType parseType(String columnName, String hiveType) {
+        LOG.info("parse column " + columnName + " with type " + hiveType);
```
Comment (morningman) [inline fe/java-udf/src/main/java/org/apache/doris/jni/vec/ColumnType.java:241]: debug

File: fe/fe-core/src/main/java/org/apache/doris/catalog/HudiUtils.java [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+                        .collect(Collectors.toList());
+                // The nullable column in hudi is the union type with schemas [null, real column type]
+                if (nonNullMembers.size() == 1) {
+                    return fromAvroHudiTypeToHiveTypeString(nonNullMembers.get(0));
+                }
+                break;
+            default:
+                break;
+        }
+        String errorMsg = String.format("Unsupported hudi %s type of column %s", avroSchema.getType().getName(),
```
[the remark this answers is not in the document]
Comment (AshinGau) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/HudiUtils.java:97]: This function generates the configuration map values to JNI reader, and return string format, not column type. `Type.UNSUPPORTED` is return in `fromAvroHudiTypeToDorisType`

File: fe/fe-core/src/main/java/org/apache/doris/catalog/external/HMSExternalTable.java [hunk 9]
```diff
@@ -367,16 +374,13 @@ public List<Column> initSchema() {
...
         List<Column> tmpSchema = Lists.newArrayListWithCapacity(hmsSchema.size());
-        for (FieldSchema field : hmsSchema) {
-            tmpSchema.add(new Column(field.getName(),
-                    HiveMetaStoreClientHelper.hiveTypeToDorisType(field.getType(),
-                            IcebergExternalTable.ICEBERG_DATETIME_SCALE_MS), true, null,
-                    true, null, field.getComment(), true, null,
-                    schema.getIndexNamed(field.getName()), null));
+        for (org.apache.avro.Schema.Field hudiField : hudiSchema.getFields()) {
+            String columnName = hudiField.name().toLowerCase(Locale.ROOT);
+            tmpSchema.add(new Column(columnName, HudiUtils.fromAvroHudiTypeToDorisType(hudiField.schema()),
```
[the remark this answers is not in the document]
Comment (AshinGau) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/external/HMSExternalTable.java:382]: `schema.getIndexNamed` return the index order of the column in hudi table, but it is not implemented in some subclasses.

File: fe/java-udf/pom.xml (hunk 3 above)
Comment (AshinGau) [reply fe/java-udf/pom.xml:123]: Both of two have the same implementation of `RecordReader` interface, and hadoop-bundle is smaller than presto-bundle.

File: fe/java-udf/pom.xml (hunk 4 above)
Comment (AshinGau) [reply fe/java-udf/pom.xml:171]: conflict with hudi-bundle

File: fe/java-udf/src/main/java/org/apache/doris/jni/utils/Utils.java (hunk 6 above)
Comment (AshinGau) [reply fe/java-udf/src/main/java/org/apache/doris/jni/utils/Utils.java:36]: depend on hadoop project, which is not in `fe-common`'s `pom.xml`.

**Comment by AshinGau** [comment]
run buildall

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**